### PR TITLE
cmd/*: print warning if formula fails to load

### PR DESCRIPTION
**Pull request by SeekingMeaning**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [ ] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew typecheck` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?
- [ ] Have you successfully run `brew man` locally and committed any changes?

-----

Commands updated so far:

- `brew info`
- `brew install`
- `brew reinstall`
- `brew search`

Before:

```console
$ brew install athenason/repo/bazel@1.2.1
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
Error: No available formula or cask with the name "athenason/repo/bazel@1.2.1".
```

After:

```console
$ brew install athenason/repo/bazel@1.2.1
==> Searching for similarly named formulae...
Warning: Failed to load formula: athenason/repo/bazel@1.2.1
athenason/repo/bazel@1.2.1: Calling depends_on :java is disabled! Use "depends_on "openjdk@11", "depends_on "openjdk@8" or "depends_on "openjdk" instead.
Please report this issue to the athenason/repo tap (not Homebrew/brew or Homebrew/core), or even better, submit a PR to fix it:
  /usr/local/Homebrew/Library/Taps/athenason/homebrew-repo/Formula/bazel@1.2.1.rb:8

Error: No similarly named formulae found.
Error: No available formula or cask with the name "athenason/repo/bazel@1.2.1".
```

<details><summary>Old PR description</summary>

This PR changes the `search_formulae` method to print a warning for each formula that fails to load. This is helpful for users in understanding why e.g. `brew install` failed for a formula that they do not know is invalid, such as in #10392. On several occasions in the recent past, I too have had to manually figure out why a formula was failing to load as the `brew` output did not say why.

Before this PR, running `brew install` on an invalid formula resulted in this output:

```console
$ brew install athenason/repo/bazel@1.2.1
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
Error: No available formula or cask with the name "athenason/repo/bazel@1.2.1".
```

```console
$ brew search /bazel/                    
==> Formulae
athenason/repo/bazel@1.2.1              bazel ✔                                 bazelisk
```

Here is the output after this PR:

```console
$ brew install athenason/repo/bazel@1.2.1
==> Searching for similarly named formulae...
Warning: Failed to load formula: athenason/repo/bazel@1.2.1
athenason/repo/bazel@1.2.1: Calling depends_on :java is disabled! Use "depends_on "openjdk@11", "depends_on "openjdk@8" or "depends_on "openjdk" instead.
Please report this issue to the athenason/repo tap (not Homebrew/brew or Homebrew/core), or even better, submit a PR to fix it:
  /usr/local/Homebrew/Library/Taps/athenason/homebrew-repo/Formula/bazel@1.2.1.rb:8

Error: No similarly named formulae found.
Error: No available formula or cask with the name "athenason/repo/bazel@1.2.1".
```

```console
$ brew search /bazel/      
Warning: Failed to load formula: athenason/repo/bazel@1.2.1
athenason/repo/bazel@1.2.1: Calling depends_on :java is disabled! Use "depends_on "openjdk@11", "depends_on "openjdk@8" or "depends_on "openjdk" instead.
Please report this issue to the athenason/repo tap (not Homebrew/brew or Homebrew/core), or even better, submit a PR to fix it:
  /usr/local/Homebrew/Library/Taps/athenason/homebrew-repo/Formula/bazel@1.2.1.rb:8

==> Formulae
athenason/repo/bazel@1.2.1              bazel ✔                                 bazelisk
```

(Does this count as `critical`?)

</details>

### CODE REVIEW

**Comment by carlocab** [comment]
Thanks for fixing this! What does this do about formulae with syntax errors in them? (e.g. `depends on` instead of `depends_on`) Currently, this also produces output like
```
$ brew install athenason/repo/bazel@1.2.1
==> Searching for similarly named formulae...
Error: No similarly named formulae found.
Error: No available formula or cask with the name "athenason/repo/bazel@1.2.1".
```

**Comment by SeekingMeaning** [comment]
No dice 😢

```rb
depends on "openjdk@11"
```

```console
$ brew reinstall bazel                    
Error: No available formula or cask with the name "bazel".
```

**Comment by SeekingMeaning** [comment]
Actually, no, `brew install` outputs a warning:

```rb
keg only :provided_by_macos
```

```console
$ brew install zlib
==> Searching for similarly named formulae...
Warning: Failed to load formula: zlib
zlib: undefined method `only' for Formulary::FormulaNamespace9fb01242de382af6fbf5ab3dc4c06975::Zlib:Class
These similarly named formulae were found:
lzlib                                                        zlib
To install one of them, run (for example):
  brew install lzlib
Error: No available formula or cask with the name "zlib".
==> Searching taps on GitHub...
This formula was found in a tap:
homebrew/portable-ruby/portable-zlib
To install it, run:
  brew install homebrew/portable-ruby/portable-zlib
```

**Comment by carlocab** [comment]
Actually, I was mistaken in https://github.com/Homebrew/brew/pull/10393#issuecomment-765128388. I think the "no available formula" error is what you see currently from a syntax error, so this PR doesn't change anything about that.

**Comment by MikeMcQuaid** [comment]
> ```
> Error: No similarly named formulae found.
> Error: No available formula or cask with the name "athenason/repo/bazel@1.2.1".
> ```

I think we could skip this output entirely, actually, given we're actually failing here.

File: Library/Homebrew/formulary.rb [hunk 1]
```diff
@@ -47,6 +47,9 @@ def self.load_formula(name, path, contents, namespace, flags:)
       mod.const_set(:BUILD_FLAGS, flags)
       mod.module_eval(contents, path)
     rescue NameError, ArgumentError, ScriptError, MethodDeprecatedError => e
+      opoo "Failed to load formula: #{name}" if !quiet || ENV["HOMEBREW_FORMULARY_DEBUG"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:50]:
```suggestion
      opoo "Failed to load formula: #{name}" if !quiet || debug?
```

File: Library/Homebrew/formulary.rb [hunk 2]
```diff
@@ -47,6 +47,9 @@ def self.load_formula(name, path, contents, namespace, flags:)
...
+      $stderr.puts e unless quiet
+      $stderr.puts caller if ENV["HOMEBREW_FORMULARY_DEBUG"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:52]:
```suggestion
      $stderr.puts caller if debug?
```

File: Library/Homebrew/formulary.rb [hunk 3]
```diff
@@ -33,7 +33,7 @@ def self.formula_class_get(path)
     cache.fetch(path)
   end
 
-  def self.load_formula(name, path, contents, namespace, flags:)
+  def self.load_formula(name, path, contents, namespace, flags:, quiet: true)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:36]: Somewhat tempting to have a `quiet?` like we already have with `debug?`

File: Library/Homebrew/formulary.rb [hunk 4]
```diff
@@ -222,7 +225,7 @@ def initialize(url)
       super formula, HOMEBREW_CACHE_FORMULA/File.basename(uri.path)
     end
 
-    def load_file(flags:)
+    def load_file(*)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:228]:
```suggestion
    def load_file(flags:, quiet: true)
```

File: Library/Homebrew/formulary.rb [hunk 5]
```diff
@@ -296,7 +299,7 @@ def get_formula(spec, alias_path: nil, force_bottle: false, flags: [])
       raise TapFormulaUnavailableError.new(tap, name), "", e.backtrace
     end
 
-    def load_file(flags:)
+    def load_file(*)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/formulary.rb:302]:
```suggestion
    def load_file(flags:, quiet: true)
```

**Comment by SeekingMeaning** [comment]
Closing in favor of https://github.com/Homebrew/brew/pull/10405

**Pull request closed without merging**